llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -9034,7 +8894,7 @@ void LoopVectorizationPlanner::adjustRecipesForReductions(
       VPRecipeBase *Cur = Worklist[I];
       for (VPUser *U : Cur->getVPSingleValue()->users()) {
         auto *UserRecipe = dyn_cast<VPRecipeBase>(U);
-        if (!UserRecipe)
+        if (!UserRecipe || UserRecipe == FinalReductionResult)
```

Comment:
Must the reduction chain now terminate at FinalReductionResult? I.e., assert that `UserRecipe` is non-null by (non-dyn) casting `U` into a recipe.